### ISSUE TITLE
TP Local Build:  Use   -DOCPN_TARGET=MSVC

### ISSUE THREAD

**Issue by rgleason** [jongough/testplugin_pi#290]
The local build creates a metadata.xml that is incorrect and unaccepted.

weather_routing_pi-1.13.52.0-msvc-x86_64-10.0.22621.tar.gz
```
<target>msvc-32</target>       <---Should be "msvc-wx32"
<build-target>msvc</build-target>
<build-gtk></build-gtk>
<target-version>10.0.22621</target-version>
<target-arch>x86_64</target-arch>
<tarball-url>

```

weather_routing_pi-1.13.52.0-msvc-x86_64-10.0.22621.tar.gz

Should be 
weather_routing_pi-1.13.52.0-msvc-wx32-x86_64-10.0.22621.tar.gz

**Comment by rgleason** [comment]
Jon, is there a way to have an option with the batch file?
bld.bat  -wx32   builds wx32 and O5.7.1?
bld.bat  builds wx31  and O5.6.2

**Comment by jongough** [comment]
I have just tested OD and testplugin and both generate the correct contents of the xml file. The filename however does not contain wx32 which I suppose it could. If you want the non-wx32 build don't specify it. For this process Appveyor builds <5.7.1 in vs2017 with wx31 and >=5.7.1 in vs2022 with wx32. Is that not working?

For local windows builds you will need to add a new entry for OCPN_TARGET of MSVC when you run the cmake program. This will then generate the correct file name (it is normally in the appveyor.yml file).

**Comment by rgleason** [comment]
No, when building locally under windows it doesn't build properly as stated above.
I am using

```
cd build
cmake -T v143 -A Win32 ..
cmake --build . --target package --config relwithdebinfo >output.txt
bash ./cloudsmith-upload.sh

```
What is   " add a new entry for OCPN_TARGET of MSVC when you run the cmake program"?

**Comment by jongough** [comment]
You have to replicate what appveyor is doing. It has an environment variable called 'OCPN_TARGET' which is used to determine, partially, the filename. So the first cmake command for vs2022 should be:
`cmake -A Win32 -DOCPN_TARGET=MSVC ..`
and for vs2017
`cmake -T v143 -DOCPN_TARGET=MSVC ..`

For me using vs2022 as a gui the build works and with a standard command prompt it works.

**Comment by rgleason** [comment]
Thanks Jon.

**Comment by rgleason** [comment]
Jon, I don't have the format right using Win 11 Bash prompt. This is what I am trying.

```
REM  For Opencpn 5.7.1 and wxWidgets-3.2.1
cd build
cmake -T v143 -A Win32 -DOCPN_TARGET=MSVC "Visual Studio 17 2022" ^
      -DCMAKE_BUILD_TYPE=RelWithDebInfo  ..
cmake --build . --target tarball --config RelWithDebInfo >output.txt
bash ./cloudsmith-upload.sh
```

This is failing

```
CMake Warning:
  Ignoring extra path from command line:

   "^"

CMake Error: The source directory "C:/Users/fcgle/source/pypilot_pi/build/Visual Studio 17 2022" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
./bld.bat: line 32: -DCMAKE_BUILD_TYPE=RelWithDebInfo: command not found
CMake Warning (dev) in cmake/PluginConfigure.cmake:
  A logical block opening on the line

    C:/Users/fcgle/source/pypilot_pi/cmake/PluginConfigure.cmake:242 (if)

  closes on the line

    C:/Users/fcgle/source/pypilot_pi/cmake/PluginConfigure.cmake:248 (endif)

  with mis-matching arguments.
Call Stack (most recent call first):
  CMakeLists.txt:148 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at CMakeLists.txt:249 (target_sources):
  Policy CMP0076 is not set: target_sources() command converts relative paths
  to absolute.  Run "cmake --help-policy CMP0076" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

  An interface source of target "pypilot_pi" has a relative path.
This warning is for project developers.  Use -Wno-dev to suppress it.

```
This is what I was using before with v5.6.2
```
REM  For Opencpn 5.6.2 and wxWidgets-3.1.2
cd build
cmake -T v141_xp ..
cmake --build . --target package --config release >output.txt
bash ./cloudsmith-upload.sh
```
And this was working fine for v5.7.1 but it does not result in the right <target>msvc-wx32<target>
so I had to go into the tar.gz file and change the metadata.xml to be correct. and then save it.

What would you suggest that I try?

**Comment by rgleason** [comment]
Trying this

```
REM  For Opencpn 5.7.1 and wxWidgets-3.2.1
cd build
cmake cmake -T v143 -DOCPN_TARGET=MSVC -DCMAKE_BUILD_TYPE=RelWithDebInfo  ..
cmake --build . --target tarball --config RelWithDebInfo >output.txt
bash ./cloudsmith-upload.sh

```

**RESULT** 
```
CMake Warning:
  Ignoring extra path from command line:

   "C:/Users/fcgle/source/pypilot_pi/build/cmake"


-- Building for: Visual Studio 17 2022
-- CMakeLists: USE_GL: ON
-- CMakeLists: Build type: RelWithDebInfo
-- The C compiler identification is MSVC 19.34.31942.0
-- The CXX compiler identification is MSVC 19.34.31942.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- PluginSetup: GIT_REPOSITORY_SERVER not found setting to: github.com
-- PluginSetup: PROJECT_VERSION: 0.40.2.0
-- PluginSetup: Pypilot Version: 0.40.2.0
-- PluginSetup: OPCN_FLATPAK:
-- PluginSetup: OCPN_FLATPAK_CONFIG: , OCPN_FLATPAK_BUILD: , UNIX:
-- PluginSetup: WX_VER: 32
-- PluginSetup: PKG_TARGET_WX_VER: -wx32, WX_VER: 32
-- PluginSetup: ARCH: x86_64
-- PluginSetup: PKG_TARGET: msvc, PKG_TARGET_VERSION: 10.0.22621
-- PluginSetup: OCPN_TARGET:
-- PluginSetup: OCPN_TARGET: MSVC
-- PluginSetup: CMAKE version: 3.25.1
-- PluginConfigure: *** Staging to build pypilot_pi ***
-- PluginConfigure: CIRCLECI: , Env CIRCLECI:
-- PluginConfigure: TRAVIS: , Env TRAVIS:
-- PluginConfigure: GIT_REPOSITORY_REMOTE: origin
-- PluginConfigure: START_URL: 19, STRING_LENGTH: 42
-- PluginConfigure: GIT_REPOSITORY: rgleason/pypilot_pi.git
-- PluginConfigure: Git Branch: "master"
-- PluginConfigure: Git Tag: ""
-- PluginConfigure: GIT_BRANCH_OR_TAG: branch
-- PluginConfigure: GIT_REPOSITORY_ITEM: master
-- PluginConfigure: CLOUDSMITH_BASE_REPOSITORY: pypilot
-- PluginConfigure: PLUGIN_EXTRA_VERSION_VARS: Not Found
-- PluginConfigure: PLUGIN_EXTRA_FORMBUILDER_HEADERS: PLUGIN_EXTRA_FORMBUILDER_HEADERS-NOTFOUND
-- PluginConfigure: PLUGIN_EXTRA_FORMBUILDER_HEADERS: Not found
-- PluginConfigure: ENV BUILD_GTK3:
-- PluginConfigure: BUILD_ENV: , BUILD_ENV_TEMP , OCPN_TARGET: MSVC
-- PluginConfigure: Doing build_gtk3:
-- PluginConfigure: PKG_TARGET_GTK:
-- PluginConfigure: ENV WX_VER: 32
-- PluginConfigure: ARCH: x86_64
-- PluginConfigure: PKG_NVR: pypilot_pi-0.40.2.0, PKG_TARGET: msvc, ARCH: x86_64, PKG_TARGET_WX_VER: -32, PKG_BUILD_GTK: , PKG_TARGET_VERSION: 10.0.22621
-- PluginConfigure: OCPN_TARGET defined: MSVC
-- PluginConfigure: PACKAGING_NAME: pypilot_pi-0.40.2.0-msvc-x86_64-32-10.0.22621-MSVC
-- PluginConfigure: PACKAGING_NAME_XML: pypilot_pi-0.40.2.0-msvc-x86_64-32-10.0.22621-MSVC
-- PluginConfigure: PKG_TARGET_FULL: msvc-32
-- PluginConfigure: PKG_BUILD_TARGET: msvc
-- PluginConfigure: PKG_BUILD_GTK:
-- PluginConfigure: PKG_BUILT_WITH_GTK:
-- PluginConfigure: *.in files generated in C:/Users/fcgle/source/pypilot_pi/build
-- PluginConfigure: DEBUG available
-- PluginConfigure: Set SNPRINTF
-- PluginConfigure: Finding package OpenGL
-- Found OpenGL: opengl32
-- PluginConfigure: Found OpenGL...
-- PluginConfigure:     Lib: opengl32glu32
-- PluginConfigure:     Include:
-- PluginConfigure:     Revised GL Lib: opengl32
-- PluginConfigure: wxWidgets components: base;core;net;xml;html;adv;aui;gl
CMake Error at C:/Program Files/CMake/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find wxWidgets (missing: wxWidgets_LIBRARIES
  wxWidgets_INCLUDE_DIRS base core net xml html adv aui gl)
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files/CMake/share/cmake-3.25/Modules/FindwxWidgets.cmake:977 (find_package_handle_standard_args)
  cmake/PluginConfigure.cmake:546 (find_package)
  CMakeLists.txt:148 (include)


-- Configuring incomplete, errors occurred!
See also "C:/Users/fcgle/source/pypilot_pi/build/CMakeFiles/CMakeOutput.log".
+ PROD_REPO=opencpn/pypilot-prod
+ BETA_REPO=opencpn/pypilot-beta
+ ALPHA_REPO=opencpn/pypilot-alpha
+ LOCAL_BUILD=false
+ '[' '' ']'
+ '[' '' ']'
+ '[' '' ']'
+ BUILD_ID=1
+ BUILD_DIR=.
+ BUILD_BRANCH=
+ BUILD_TAG=
+ PKG_EXT=deb
+ LOCAL_BUILD=true
+ set +x
+ '[' '' ']'
+ '[' '' ']'
++ git rev-parse --short=7 HEAD
+ commit=d816c08
++ git tag --contains HEAD
+ tag=
+ ls -la
total 66
drwxr-xr-x 1 fcgle 197609     0 Mar  4 14:37 .
drwxr-xr-x 1 fcgle 197609     0 Mar  4 14:24 ..
-rw-r--r-- 1 fcgle 197609 22539 Mar  4 14:37 CMakeCache.txt
drwxr-xr-x 1 fcgle 197609     0 Mar  4 14:37 CMakeFiles
-rw-r--r-- 1 fcgle 197609   741 Mar  4 14:37 PluginCPackOptions.cmake
-rwxr-xr-x 1 fcgle 197609  8831 Mar  4 14:37 cloudsmith-upload.sh
-rw-r--r-- 1 fcgle 197609   133 Mar  4 14:37 output.txt
-rw-r--r-- 1 fcgle 197609   311 Mar  4 14:37 pkg_version.sh
-rw-r--r-- 1 fcgle 197609  1197 Mar  4 14:37 pypilot_pi-0.40.2.0-msvc-x86_64-32-10.0.22621-MSVC.xml
+ pwd
/c/Users/fcgle/source/pypilot_pi/build
++ ls ./pypilot_pi-0.40.2.0-msvc-x86_64-32-10.0.22621-MSVC.xml
+ xml=./pypilot_pi-0.40.2.0-msvc-x86_64-32-10.0.22621-MSVC.xml
+ cat ./pypilot_pi-0.40.2.0-msvc-x86_64-32-10.0.22621-MSVC.xml
<?xml version="1.0" encoding="UTF-8"?>
<plugin version="1">
  <name> pypilot </name>
  <version> 0.40.2.0 </version>
  <release> 0 </release>
  <summary> Pypilot PlugIn for OpenCPN </summary>

  <api-version> 1.16 </api-version>
  <open-source> yes </open-source>
  <author> Sean D'Epagnier </author>
  <source> https://github.com/rgleason/pypilot_pi.git </source>

  <description>
    Control the free software autopilot pypilot. See http://pypilot.org for more details. /n /;The plugin connects to pypilot directly implementing a control interface to configure, /n /;calibrate and command pypilot from opencpn. For more control and tuning route-following /n /;logic, consider the autopilot route plugin. /n /
  </description>

  **<target>msvc-32</target>**    **<<<<----LOOK AT THIS**
  <build-target>msvc</build-target>
  <build-gtk></build-gtk>
  <target-version>10.0.22621</target-version>
  <target-arch>x86_64</target-arch>
  <tarball-url>
    https://dl.cloudsmith.io/public/--pkg_repo--/raw/names/--name--/versions/--version--/--filename--
  </tarball-url>
  <info-url> https://opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:opencpn_user_manual:plugins:safety:odraw1.6_pi </info-url>
</plugin>
++ ls './*.tar.gz'
ls: cannot access './*.tar.gz': No such file or directory
+ tarball=

```

**Comment by rgleason** [comment]
Actually, this is what builds the 3 files including tar.gz

```
cd build
cmake -T v143 -A Win32 ..
cmake --build . --target package --config relwithdebinfo >output.txt
bash ./cloudsmith-upload.sh

```

The problem is that the metadata.xml included in the tar.gz has the wrong target and I have to edit it each time I want to test locally.

It shows as <target>msvc-32</target>  so I have to open it and edit and save as  <target>msvc-wx32</target>

You advised that editing the file to make it come out properly was not the right way to do it. (I was the one who did that!) 
But it certainly isn't working this way!   So exactly what is the correct way to fix this?

**Comment by jongough** [comment]
As far as I can tell the automated build is now working for all environments. I don't have a win11 to test with. This means the definitive answer for how to build is contained in either the circleci stream or the appveyor stream. Currently appveyor builds with vs2017 for current releases and vs2022 for the beta and future releases. The appveyor.yml file shows what is needed for each environment and the commands that have to be issued.

If you are building for the beta the means you need vs2022 installed on the machine you are building with and you need to have installed the dependencies, `poedit`, `git` and `wget`. you also need installed `wxMSW-3.2.1_cx14x_Dev` and `wxWidgets-3.2.1-headers`. Within the directory of the project you need to download `PVW32Con.exe` and within the build directory have `opencpn.lib` and within the buildwin directory the unzipped `OpenCPN_buildwin-4.99a.7z`.

Now you need the other environment variables setup:
```
WX_VER=32
wxWidgets_INCLUDE_DIRS=C:\wxWidgets-3.2.1\include
wxWidgets_LIBRARIES=C:\wxWidgets-3.2.1\lib\vc14x_dll
wxWidgets_LIB_DIR=C:\wxWidgets-3.2.1\lib\vx14x_dll
wxWIDGETS_ROOT_DIR=C:\wxWidgets-3.2.1
WXWIN=C:\wxWidgets-3.2.1
```
These will need to be adjusted to where you have installed the wx32 stuff. If you issue `set` from a standard command prompt and see these all is OK, if not you need to ensure they are provided. My Win10 environment has multiple versions of wxWidgets installed so I can build for the needed environments. 

Now you would need to issue `cmake -A Win32 -G "Visual Studio 17 2023" -DOCPN_TARGET=MSVC ..` 
Then when that has completed you can issue `cmake --build . --target package --config RelWithDebInfo`

Things are not as simple as they used to be with all the different environments that have to be catered for now.

For my purposes I normally have the cmake gui installed and the VS2017 & VS2022 installed and do the builds. Using the cmake gui to create all the solution and project files needed then build OCPN and add the OD or Testplugin project file to the OCPN solution and build them, then move the resulting dll's to the correct location so I can debug the whole lot.

**Comment by rgleason** [comment]
This is a good description for developers to use to setup. It is what I have already done and have been building for quite awhile. I have both wxWidgets 3.2.1 and wxWidgets 3.1.2 installed under fcgle/sources where I keep all the git repos and plugins.

![Env](https://user-images.githubusercontent.com/2363959/222933460-9d28eaf5-eac5-4b12-a542-adec4f793f67.png)

Note that by using  %WXWIN% in all the variables, it is easy to redirect to the other wxWidgets.


I will try to follow your suggestions. The plugins build with the cmake commands I showed above, but don't seem to be building with your suggestion (yet)

- cmake -A Win32 -G "Visual Studio 17 2023" -DOCPN_TARGET=MSVC ..
- cmake --build . --target package --config RelWithDebInfo

**Comment by jongough** [comment]
I have just merged updates to master and the new version is v1.0.230.0. This should work for all platforms. I have used the commands, but they do have to be done from a standard windows cmd prompt not a developers cmd prompt. Not sure why, but VS seems to muck around with something and some builds work and then the same build fails on the second go. The standard cmd prompt does not seem to show the same issue.
